staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go

```diff
@@ -0,0 +1,316 @@
+	curState atomic.Value
+}
+
+// TypedConfigObjectReference is a reference to a relevant config API object.
+// No namespace is needed because none of these objects is namespaced.
+type TypedConfigObjectReference struct {
+	Kind string
+	Name string
+}
+
+func (tr *TypedConfigObjectReference) String() string {
+	return tr.Kind + "/" + tr.Name
+}
+
+// rmSetup is invoked at startup to create the infrastructure of this filter
+func rmSetup(kubeClient kubernetes.Interface, serverConcurrencyLimit int, requestWaitLimit time.Duration, clk clock.Clock) *requestManagement {
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+	fci := kubeInformerFactory.Flowcontrol().V1alpha1()
+	pli := fci.PriorityLevelConfigurations()
+	fsi := fci.FlowSchemas()
+	reqMgmt := &requestManagement{
+		clk:                    clk,
+		configQueue:            workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour), "req_mgmt_config_queue"),
+		plInformer:             pli.Informer(),
+		plLister:               pli.Lister(),
+		fsInformer:             fsi.Informer(),
+		fsLister:               fsi.Lister(),
+		serverConcurrencyLimit: serverConcurrencyLimit,
+		requestWaitLimit:       requestWaitLimit,
+	}
+	// TODO: finish implementation
+	return reqMgmt
+}
+
+// WithRequestManagement limits the number of in-flight requests in a fine-grained way
+func WithRequestManagement(
+	handler http.Handler,
+	clientConfig *restclient.Config,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+	longRunningRequestCheck apirequest.LongRunningRequestCheck,
+) http.Handler {
+	kubeClient, err := kubernetes.NewForConfig(clientConfig)
+	if err != nil {
+		klog.Errorf("Failed to construct Kubernetes client (%s), skipping prioritization and fairness filter", err.Error())
+		return handler
+	}
+	return WithRequestManagementByClient(handler, kubeClient, serverConcurrencyLimit, requestWaitLimit, longRunningRequestCheck, clock.RealClock{})
+}
+
+// WithRequestManagementByClient limits the number of in-flight
+// requests in a fine-grained way and is more appropriate than
+// WithRequestManagement for testing
+func WithRequestManagementByClient(
+	handler http.Handler,
+	kubeClient kubernetes.Interface,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+	longRunningRequestCheck apirequest.LongRunningRequestCheck,
+	clk clock.Clock,
+) http.Handler {
+	reqMgmt := rmSetup(kubeClient, serverConcurrencyLimit, requestWaitLimit, clk)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no RequestInfo found in context, handler chain must be wrong"))
+			return
+		}
+
+		// Skip tracking long running events.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		for {
```

Comment:
I'm not sure I understand the purpose of the `for` loop here.  Additionally I'm not sure I understand from the skeleton how code exits this loop.

Reply:
I think we might want to 'return' after the select cases

Reply:
Oops!  Right!
The purpose of the loop was just to be able to do a retry in the quiescing case.